### PR TITLE
Anna/Volunteer Sign-Up

### PR DESCRIPTION
**Pull request by alin1214**
# Pull Request

## Description

Volunteers are able to see available shifts and be able to sign up based on availability. They are able to sign up as a lead driver or just a general volunteer and drop shifts that they are no longer available for. 

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] Performance improvement
- [ ] Other (please describe):

## Related Issues (put task name here from notion)

Volunteer Shift Sign-up (Mobile/Web)

### Code Review Preparation
- [x] I have performed a self-review and reviewed the changed files
- [x] My code follows the project's style guidelines and passes linting
- [x] I have added comments in hard-to-understand areas and updated documentation

### Testing (Ignore for the first PR)
- [ ] I have added relevant tests and they pass locally
- [ ] I have tested the changes manually, including edge cases

### Code Quality & Security
- [ ] I have formatted my code and removed debugging/unused code
- [x] No TypeScript errors or linting issues
- [ ] No sensitive information committed (API keys, passwords, etc.)
- [ ] User inputs are validated and sanitized

### Dependencies & Deployment
- [ ] Dependencies are documented and package-lock.json is updated if needed
- [ ] Changes work with Firebase emulator and environment variables are documentedRetry

## Screenshots (If it is a front end feature screenshot is required)

Screenshot 2026-03-25 at 4 29 07 AM
Screenshot 2026-03-25 at 4 29 42 AM
Screenshot 2026-03-25 at 4 28 53 AM

## Summary by CodeRabbit

* **New Features**
  * Volunteer shift sign-up page with filters for shift type and availability
  * Responsive shift list (mobile & desktop) showing grouped shifts and capacities
  * Sign up/drop flows with confirmation dialogs and selectable volunteer role
  * Protected volunteer-only route and "Shifts" side navigation entry
  * Reusable confirmation modal that prevents background scrolling while open

* **Style**
  * Tightened multiselect trigger button height for a more compact layout

### CODE REVIEW

File: components/schedule/ShiftListView.tsx [hunk 1]
```diff
@@ -0,0 +1,484 @@
+"use client";
+
+import { useState } from "react";
+import { Timestamp } from "firebase/firestore";
+import { TimeBlock } from "@/types/schedule";
+import ConfirmModal from "@/components/general/ConfirmModal";
+import Button from "@/components/form/Button";
+import { useTimeBlocks } from "@/lib/queries/timeblocks";
+
+type TBWithTypes = TimeBlock & {
```
Comment (constrictingsnake) [inline components/schedule/ShiftListView.tsx]: TBWithTypes doesn't need to exist, all the data can be derived from the current timeblock already

File: components/schedule/ShiftListView.tsx [hunk 2]
```diff
@@ -0,0 +1,484 @@
...
+                                        .toLocaleDateString("en-US", {
+                                            weekday: "short",
+                                        })
+                                        .toUpperCase()}`}
+                                </div>
+                            </div>
+
+                            {/* shifts */}
+                            {blocks.map((tb) => {
+                                const tbWithTypes = tb as TBWithTypes;
```
Comment (constrictingsnake) [inline components/schedule/ShiftListView.tsx]: TBWithTypes not necessary

File: components/schedule/ShiftListView.tsx [hunk 3]
```diff
@@ -0,0 +1,484 @@
...
+                                            weekday: "short",
+                                        })
+                                        .toUpperCase()}`}
+                                </div>
+                            </div>
+
+                            {/* shifts */}
+                            <div className="flex flex-col gap-3 flex-1">
+                                {blocks.map((tb) => {
+                                    const tbWithTypes = tb as TBWithTypes;
```
Comment (constrictingsnake) [inline components/schedule/ShiftListView.tsx]: TBWithTypes not necessary

File: components/schedule/ShiftListView.tsx [hunk 4]
```diff
@@ -0,0 +1,484 @@
...
+                                )}-${formatTime(tb.endTime)}`;
+
+                                const isSignedUp = tb.volunteerGroups.some(
+                                    (group) =>
+                                        group.volunterIDs.includes(
+                                            currentUserID
+                                        )
+                                );
+
+                                const type =
```
Comment (constrictingsnake) [inline components/schedule/ShiftListView.tsx]: get type from the type field on the timeblock

File: components/schedule/ShiftListView.tsx [hunk 5]
```diff
@@ -0,0 +1,484 @@
...
+                                            )
+                                    );
+
+                                    const isFull = tb.volunteerGroups.every(
+                                        (group) =>
+                                            group.volunterIDs.length >=
+                                            group.maxNum
+                                    );
+
+                                    const type =
```
Comment (constrictingsnake) [inline components/schedule/ShiftListView.tsx]: get type from the type field on the timeblock

File: components/schedule/ShiftListView.tsx [hunk 6]
```diff
@@ -0,0 +1,484 @@
...
+                                            key={tb.id}
+                                            className="flex items-center justify-between"
+                                        >
+                                            <div className="flex items-center gap-4">
+                                                {/* time dots */}
+                                                <div
+                                                    className={`w-3 h-3 rounded-full ${
+                                                        type === "Warehouse"
+                                                            ? "bg-yellow-400"
+                                                            : "bg-teal-500"
```
Comment (constrictingsnake) [inline components/schedule/ShiftListView.tsx]: the color for the pickup/delivery shifts is wrong

File: components/schedule/ShiftListView.tsx [hunk 7]
```diff
@@ -0,0 +1,484 @@
...
+                                            </div>
+
+                                            <div className="flex flex-col items-end gap-2">
+                                                {/* time */}
+                                                <div className="flex items-center gap-2 text-xs">
+                                                    <div
+                                                        className={`w-3 h-3 rounded-full ${
+                                                            type === "Warehouse"
+                                                                ? "bg-yellow-400"
+                                                                : "bg-teal-500"
```
Comment (constrictingsnake) [inline components/schedule/ShiftListView.tsx]: the color for the pickup/delivery shifts doesn't match the figma

File: components/schedule/ShiftListView.tsx [hunk 8]
```diff
@@ -0,0 +1,484 @@
...
+    timeBlocks: TimeBlock[];
+    currentUserID: string;
+};
+
+export default function ShiftListView({ timeBlocks, currentUserID }: Props) {
+    const [selectedTB, setSelectedTB] = useState<TimeBlock | null>(null);
+    const [action, setAction] = useState<"signup" | "drop" | null>(null);
+    const { setTimeblockToast } = useTimeBlocks();
+
+    const [volunteerType, setVolunteerType] = useState<
```
Comment (constrictingsnake) [inline components/schedule/ShiftListView.tsx]: volunteer types are already accessible through the volunteerGroup array within a timeblock, where the volunteerGroup has a name which would be the names of each of the types of volunteers

File: components/schedule/ShiftListView.tsx [hunk 9]
```diff
@@ -0,0 +1,484 @@
...
+                                        )
+                                );
+
+                                const type =
+                                    tb.tasks.length > 0 &&
+                                    "donor" in tb.tasks[0]
+                                        ? "Pickups / Deliveries"
+                                        : "Warehouse";
+
+                                const leadDrivers = Object.values(
```
Comment (constrictingsnake) [inline components/schedule/ShiftListView.tsx]: the front end should be able to abstract whatever volunteer groups are within the time block, stuff like leadDrivers and generalVolunteers doesn't need to be hardcoded

File: components/schedule/ShiftListView.tsx [hunk 10]
```diff
@@ -0,0 +1,484 @@
...
+                                    tbWithTypes.volunteerTypes || {}
+                                ).filter(
+                                    (type) => type === "Lead driver"
+                                ).length;
+
+                                const generalVolunteers = Object.values(
```
Comment (constrictingsnake) [inline components/schedule/ShiftListView.tsx]: same as above, should be abstracted

File: components/schedule/ShiftListView.tsx [hunk 11]
```diff
@@ -0,0 +1,484 @@
...
+                                        <div className="flex justify-between items-start">
+                                            <div className="flex flex-col gap-1">
+                                                <div className="text-sm text-gray-700">
+                                                    {type}
+                                                </div>
+
+                                                <div className="text-primary text-xs">
+                                                    <div>
+                                                        {/* assuming there's only two lead drivers */}
+                                                        {leadDrivers}/2 lead
```
Comment (constrictingsnake) [inline components/schedule/ShiftListView.tsx]: It's not needed to assume the max num of volunteers for anything as the max number is present in the VolunteerGroup array

File: components/schedule/ShiftListView.tsx [hunk 12]
```diff
@@ -0,0 +1,484 @@
...
+                                        .toUpperCase()}`}
+                                </div>
+                            </div>
+
+                            {/* shifts */}
+                            <div className="flex flex-col gap-3 flex-1">
+                                {blocks.map((tb) => {
+                                    const tbWithTypes = tb as TBWithTypes;
+
+                                    const totalGeneralCapacity =
```
Comment (constrictingsnake) [inline components/schedule/ShiftListView.tsx]: use maxNum field of the volunteerGroups

File: components/schedule/ShiftListView.tsx [hunk 13]
```diff
@@ -0,0 +1,484 @@
...
+                                            group.maxNum
+                                    );
+
+                                    const type =
+                                        tb.tasks.length > 0 &&
+                                        "donor" in tb.tasks[0]
+                                            ? "Pickups / Deliveries"
+                                            : "Warehouse";
+
+                                    const leadDrivers = Object.values(
```
Comment (constrictingsnake) [inline components/schedule/ShiftListView.tsx]: refer to previous comment about abstraction

File: components/schedule/ShiftListView.tsx [hunk 14]
```diff
@@ -0,0 +1,484 @@
...
+                                        tbWithTypes.volunteerTypes || {}
+                                    ).filter(
+                                        (type) => type === "Lead driver"
+                                    ).length;
+
+                                    const generalVolunteers = Object.values(
```
Comment (constrictingsnake) [inline components/schedule/ShiftListView.tsx]: refer to previous comment about abstraction

File: components/schedule/ShiftListView.tsx [hunk 15]
```diff
@@ -0,0 +1,484 @@
...
+
+                                                <div className="w-36 text-gray-700">
+                                                    {type}
+                                                </div>
+
+                                                {/* volunteets */}
+                                                <div className="text-primary px-5">
+                                                    <div className="flex flex-col text-primary text-sm">
+                                                        <span>
+                                                            {leadDrivers}/2 lead
```
Comment (constrictingsnake) [inline components/schedule/ShiftListView.tsx]: same comment as before with the leadDrivers not needing to be assumed

File: components/schedule/ShiftListView.tsx [hunk 16]
```diff
@@ -0,0 +1,484 @@
...
+            {selectedTB && action === "signup" && (
+                <ConfirmModal
+                    title="Confirm Sign Up"
+                    confirmLabel="Sign up"
+                    onCancel={() => {
+                        setSelectedTB(null);
+                        setAction(null);
+                        setVolunteerType(null);
+                    }}
+                    onConfirm={() => {
```
Comment (constrictingsnake) [inline components/schedule/ShiftListView.tsx:301]: within this confirmation, what gets passed into the mutation is your local volunteer types field, which gets uploaded to firestore and that's the only place the user's id is stored. as such, when elsewhere in your code you see if the user signed up for anything, you check volunteerGroups (where the user's id should be within the volunteerIDs array) but since its never stored there, its impossible to drop a shift. modify the confirm modal to add the volunteer's uid to the volunteerIDs array for the respective type of shift that they sign up for

File: components/schedule/ShiftListView.tsx [hunk 17]
```diff
@@ -0,0 +1,484 @@
...
+                        </label>
+                    </div>
+                    <p className="mt-5">
+                        Are you sure you are available this date?
+                    </p>
+                </ConfirmModal>
+            )}
+
+            {/* drop model */}
+            {selectedTB && action === "drop" && (
```
Comment (constrictingsnake) [inline components/schedule/ShiftListView.tsx:314]: modify the drop modal the same way youd have to modify the confirm modal to get rid of the specific user's id from the volunteerIDs array

File: components/general/ConfirmModal.tsx [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+}: ConfirmModalProps) {
+    // prevents background scroll when modal is open
+    useEffect(() => {
+        document.body.style.overflow = "hidden";
+        return () => {
+            document.body.style.overflow = "";
+        };
+    }, []);
+
+    return createPortal(
```
Comment (constrictingsnake) [inline components/general/ConfirmModal.tsx]: In general, align the modal to match the figma more closely, like with making sure the right text is there and the spacing and stuff

File: app/volunteer-signup/page.tsx [hunk 19]
```diff
@@ -0,0 +1,98 @@
...
+            </div>
+        );
+    }
+
+    return (
+        <div className="flex flex-col h-full">
+            <div className="bg-white rounded-l px-6 py-1">
+                {/* shift type + spots available dropdwon */}
+                <div className="flex items-center gap-4 py-3 border-b border-gray-200">
+                    <DropdownMultiselect
```
Comment (constrictingsnake) [inline app/volunteer-signup/page.tsx:92]: for both of these dropdown multiselects, they should align with the figma more

File: app/volunteer-signup/page.tsx [hunk 20]
```diff
@@ -0,0 +1,98 @@
...
+        "Pickups / Deliveries",
+    ]);
+
+    const [selectedAvailability, setSelectedAvailability] = useState<string[]>([
+        "Available",
+        "Full",
+    ]);
+
+    const filteredTimeBlocks = timeBlocks.filter((tb) => {
+        const type =
```
Comment (constrictingsnake) [inline app/volunteer-signup/page.tsx:42]: type is in the type field of the timeblocks

File: app/volunteer-signup/layout.tsx [hunk 21]
```diff
@@ -0,0 +1,71 @@
...
+                                >
+                                    ☰
+                                </button>
+
+                                <h1 className="text-[24px] font-bold text-primary">
+                                    Shift Sign-up
+                                </h1>
+                            </div>
+
+                            <button className="hidden md:flex mt-1 md:mt-4 text-sm border-b-2 border-primary text-primary">
```
Comment (constrictingsnake) [inline app/volunteer-signup/layout.tsx]: list view is the only view that we're gonna have for this, so this isn't needed

File: app/volunteer-signup/layout.tsx [hunk 22]
```diff
@@ -0,0 +1,71 @@
...
+                        <SideNavbar />
+                    </div>
+
+                    <div className="flex-1 md:bg-gray-100 py-4 pr-4 md:pl-10 md:pr-6 flex flex-col">
+                        
+                        {/* header section */}
+                        <div className="border-b border-gray-200 md:border-none pb-3 ml-4 mt-2 md:ml-2">
+                            
+                            <div className="flex items-center gap-3">
+                                <button
```
Comment (constrictingsnake) [inline app/volunteer-signup/layout.tsx]: implementing the mobile version of the navbar is outside the scope of the task,  you can remove this

File: app/volunteer-signup/layout.tsx [hunk 23]
```diff
@@ -0,0 +1,71 @@
...
+    const [isSidebarOpen, setIsSidebarOpen] = useState(false);
+
+    return (
+        <ProtectedRoute allow={["Volunteer"]}>
+            <div className="h-full w-full flex flex-col font-family-roboto">
+                {/* main content */}
+                <div className="flex flex-1 relative">
+
+                    {/* mobile only */}
+                    {isSidebarOpen && (
```
Comment (constrictingsnake) [inline app/volunteer-signup/layout.tsx]: mobile navbar portion not necessary; you can remove this portion

File: components/schedule/ShiftListView.tsx [hunk 24]
```diff
@@ -0,0 +1,484 @@
...
+                                ).filter(
+                                    (type) => type === "General volunteer"
+                                ).length;
+
+                                return (
+                                    <div key={tb.id} className="pb-4">
+                                        <div className="flex justify-between items-start">
+                                            <div className="flex flex-col gap-1">
+                                                <div className="text-sm text-gray-700">
+                                                    {type}
```
Comment (constrictingsnake) [inline components/schedule/ShiftListView.tsx]: rather than type, it would be better to display the title of the shift

File: components/schedule/ShiftListView.tsx [hunk 25]
```diff
@@ -0,0 +1,484 @@
...
+                                                <div className="w-30 text-gray-700">
+                                                    {timeRange}
+                                                </div>
+
+                                                <div className="w-36 text-gray-700">
+                                                    {type}
+                                                </div>
+
+                                                {/* volunteets */}
+                                                <div className="text-primary px-5">
```
Comment (constrictingsnake) [inline components/schedule/ShiftListView.tsx]: same comment as above, Line 122

**Review by constrictingsnake — CHANGES_REQUESTED** [review]

File: app/volunteer-signup/page.tsx (hunk 19 above)
Comment (joelchem) [reply app/volunteer-signup/page.tsx:92]: copy styling from user management here for the dropdowns to stop them from looking a little squished. it could also be because of the styling of the rest of the page

**Review by joelchem — APPROVED** [review]
